Make the mixer robust to adapter panics and end processing early on timeout

ZackButcher opened the pull request:
Wrap the dispatch to our adapters in a function which will recover them panicking. This wrapper also polls for the cancellation of the request's context and aborts early if it is canceled (e.g. because we exceeded the requests's deadline).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/istio/mixer/177)
<!-- Reviewable:end -->

ZackButcher commented:
I verified this by both having the manager panic and adding a deadline flag to mixc. In both cases we got the expected behavior. If we're interested in a flag for setting the request deadline in mixc I can include it in this PR too.

ZackButcher commented on pkg/api/methodHandlers.go:
```diff
@@ -131,6 +138,38 @@ func (h *methodHandlers) execute(ctx context.Context, tracker attribute.Tracker,
+		c <- result{o, err}
+	}()
+
+	select {
```
Actually, giving this more thought: if we keep the timeout check above in `execute` we can remove the check here and not spin up a new goroutine for the adapter, keeping it all in the request thread. I *think* that'd be better, thoughts?

douglas-reid commented:
Missing tests?

geeknoid commented on pkg/api/methodHandlers.go:
```diff
@@ -131,6 +138,38 @@ func (h *methodHandlers) execute(ctx context.Context, tracker attribute.Tracker,
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				glog.Warningf("Recovered from panic in adapter '%s'", conf.Adapter.Name)
```
Should log this as an error.

Should disable/remove the aspect once it has failed so we don't end up continuously failing. Maybe we should keep a counter and allow N failures per minute or something....

ZackButcher commented:
There's no tests for the entire file, I can start some.

ZackButcher commented:
Updated to keep things in the request thread; the semantics are essentially unchanged but we don't have to run the adapter in a different thread. Do we like this version more or less than the other?

geeknoid commented:
We ultimately need to imbue the code with parallelism. At some point, we
need a goroutine pool that will take requests from the front-end and issue
the calls into the adapters. The adapters will block on I/O, but we can
keep pumping incoming requests (until some predetermined limit at which
point we need to start failing additional requests)

I think the old mixologist code already has this kind of model.

Given this, seems like it might be time to start embracing go routines and
not shy away...

On Fri, Jan 20, 2017 at 3:58 PM, Zack <notifications@github.com> wrote:

> Updated to keep things in the request thread; the semantics are
> essentially unchanged but we don't have to run the adapter in a different
> thread. Do we like this version more or less than the other?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/mixer/pull/177#issuecomment-274211373>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AVucHcpPZPsVdzmPXlDYMt1Tcz8Z6mvqks5rUUomgaJpZM4Lp3Ka>
> .
>

douglas-reid commented:
Yeah, the prototype used go routines. We learned some lessons in this regard. @mandarjog can give you a full dump of the details (debugging, performance troubleshooting, etc.).

ZackButcher commented:
The version previously written had no parallelism of adapter execution, the only parallelism was to enable monitoring both the context's cancellation and the execution of the adapter. The current version is a little worse in that we wait for the adapter to finish executing before returning a `DEADLINE_EXCEEDED` to the user, where previously we returned it immediately, but there's no extra expenditure of resources since we can't kill the goroutine executing the adapter anyway.

It's as easy to implement parallel execution of adapters with the code as currently written compared to how it was previously written (changing `h.execWrapper(ctx, conf, ab)` to `go h.execWrapper(ctx, conf, ab)`) but I think the current version is easier to understand.

mandarjog commented on pkg/api/methodHandlers.go:
```diff
@@ -131,6 +136,24 @@ func (h *methodHandlers) execute(ctx context.Context, tracker attribute.Tracker,
+			if panicErr, ok := r.(error); ok {
+				err = panicErr
+			} else {
+				err = fmt.Errorf("adapter '%s' panicked", conf.Adapter.Name)
```
we still want recovers return value captured in the error.

```go
fmt.Errorf("adapter '%s' panicked with '%v'", conf.Adapter.Name, r)
```

mandarjog commented on pkg/api/methodHandlers.go:
```diff
@@ -116,9 +116,14 @@ func (h *methodHandlers) execute(ctx context.Context, tracker attribute.Tracker,
-		out, err := h.mngr.Execute(conf, ab, h.eval)
+		select {
+		case <-ctx.Done():
+			return newStatusWithMessage(code.Code_DEADLINE_EXCEEDED, ctx.Err().Error())
```
Would be nice to send back results available so far...

ZackButcher replied:
I'm not sure we want to do that. For example, what if my auth adapter takes too long but would've denied the request, and all of the other adapters I executed gave an OK result? I don't think we want to return an OK in that case, right?

There's also another problem with that's written: that it's possible for Done to be closed for a reason other than deadline exceeded (e.g. callers cancels the request).

ZackButcher commented:
Thought about this over the weekend: I think this is better if we push catching adapter panics down into the manager, so that manager.Execute always returns either output or an err. I think this will make things cleaner when we move to concurrent adapter execution.

I also updated the panic-handling code to always return the recover, even if its not an error, by wrapping it in another error providing context on the failure.

Still open is how to handle the context finishing before we're done executing adapters. As written, the loop in methodHandlers exit as soon as an adapter returns a not-OK code. Mandar suggested returning the output we have to far, but I'm not sure we want to return an OK if the request is canceled or we deadline exceed. Added a TODO to look into this behavior later.

mandarjog commented on pkg/adapterManager/manager.go:
```diff
@@ -96,11 +96,17 @@ func (m *Manager) Execute(cfg *aspect.CombinedConfig, attrs attribute.Bag, mappe
 		return nil, err
 	}
 
+	defer func() {
```
We should wrap *all* AdapterImpl code in a try-catch.

asp.Execute and  m.CacheGet both call AdapterImpl supplied code. 
pl. move defer before the cacheGet call.

ZackButcher commented on pkg/adapterManager/manager_test.go:
```diff
@@ -38,12 +43,51 @@ type (
 	return m.kind
 }
 
+func newTestManager(name string, throwOnNewAspect bool, aspectThrowOnExecute bool) testManager {
```
Do we have a manager/aspect for testing already? If not, we should create one (I'll need to another to use in a different package as part of testing the pkg/api surface in a PR coming soon).

douglas-reid commented on pkg/adapterManager/manager_test.go:
```diff
@@ -60,3 +104,47 @@ func TestManager(t *testing.T) {
+		&istioconfig.Adapter{Name: name},
+	}
+
+	if _, err := m.Execute(cfg, nil, nil); err != nil {
```
might consider re-ordering this:

```
_, err := m.Execute...
if err == nil {
   t.Error("Execute(): expected error")
}
if !strings... {
    t.Errorf("Execute() => got '%v', want message containing 'NewAspect'", err)
}
```

That would eliminate the nested ifs.

Same suggestion for `TestRecovery_AspectExecute`.

douglas-reid commented on pkg/api/methodHandlers_test.go:
```diff
@@ -0,0 +1,57 @@
+
+	s := handler.execute(ctx, attribute.NewManager().NewTracker(), &mixerpb.Attributes{}, Check)
+	if s.Code != int32(code.Code_DEADLINE_EXCEEDED) {
+		t.Errorf("Didn't return deadline exceeded when handler.execute was called with a canceled context, got status: %v", s)
```
super nit: reword error message to something like:

`"execute(canceledContext) => %v, wanted status with code %v", s,  code.Code_DEADLINE_EXCEEDED"`
